src/Orleans/Core/GrainAttributes.cs

```diff
@@ -186,6 +186,27 @@ public sealed class RandomPlacementAttribute : PlacementAttribute
         }
 
         /// <summary>
+        /// Marks a grain class as using the <c>HashBasedPlacement</c> policy.
+        /// </summary>
+        [AttributeUsage(AttributeTargets.Class, AllowMultiple = false)]
+        public sealed class HashBasedPlacementAttribute : PlacementAttribute
+        {
+            // for testing
+            internal bool SortSiloList { get; }
+
+            public HashBasedPlacementAttribute() :
+                base(HashBasedPlacement.Singleton)
+            {}
+
+            // for testing
+            internal HashBasedPlacementAttribute(bool sortSiloList = false)
```

Comment:
If silos are not sorted, each silo may determine that grain X belongs to a different silo, which in my view goes against the idea of using hash-based placement (may as well use random). Or is there a natural sort order which all silos converge on but which we cannot easily determine in our tests?

I believe that the placement used here should leverage the consistent ring balancer so that placement decisions converge and so that each grain is collocated with any grain services responsible for it.